lib/iex/lib/iex/evaluator.ex

```diff
@@ -83,20 +83,38 @@ defmodule IEx.Evaluator do
     file = Keyword.get(opts, :file, "nofile")
     line = Keyword.get(opts, :line, 1)
     column = Keyword.get(opts, :column, 1)
+    last_op = Keyword.get(opts, :last_op, :none)
     charlist = String.to_charlist(input)
 
     result =
-      with {:ok, tokens} <- :elixir.string_to_tokens(charlist, line, column, file, opts) do
-        :elixir.tokens_to_quoted(adjust_operator(tokens, line, column, file, opts), file, opts)
+      with {:ok, tokens} <- :elixir.string_to_tokens(charlist, line, column, file, opts),
+           {:ok, adjusted_tokens} <- adjust_operator(tokens, line, column, file, opts, last_op),
+           {:ok, forms} <- :elixir.tokens_to_quoted(adjusted_tokens, file, opts) do
+        last_op =
+          Enum.reduce_while(tokens, :other, fn
+            {:match_op, _, _}, _ -> {:halt, :match}
+            _, acc -> {:cont, acc}
+          end)
+
+        {:ok, forms, last_op}
       end
 
     case result do
-      {:ok, forms} ->
-        {:ok, forms, ""}
+      {:ok, forms, last_op} ->
+        {:ok, forms, "", last_op}
 
       {:error, {_, _, ""}} ->
         {:incomplete, input}
 
+      {:error, {location, :malformed_pipe, token}} ->
+        :elixir_errors.parse_error(
+          location,
+          file,
+          "pipe shorthand is not allowed immediately after a match expression in IEx; to make it work, surround the whole pipeline with parentheses ",
+          token,
+          {charlist, line, column}
+        )
```

Comment:
The error message is good but I think we should already return the `{location, message, token}` format from `adjust_operator` and then it falls under the clause below.

Reply:
The only change to the error message is to remove `;`. We rarely use it in Elixir error messages. :)

Reply:
Only this comment left and we should be good. In a nutshell, I think you should return the message above instead of :malformed_pipe in `adjust_operator`.

Reply:
@josevalim ✓ Done.